l10n_be_cooperator_portal_national_number/controllers/portal.py

```diff
@@ -0,0 +1,61 @@
+# SPDX-FileCopyrightText: 2024 Coop IT Easy SC
+#
+# SPDX-License-Identifier: AGPL-3.0-or-later
+
+from collections import namedtuple
+
+from odoo import _
+from odoo.http import request, route
+
+from odoo.addons.account.controllers.portal import PortalAccount
+
+
+class CooperatorPortal(PortalAccount):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        # See cooportal_portal/controllers/portal.py for why this code is the
+        # way it is.
+        if "OPTIONAL_BILLING_FIELDS" not in vars(self):
+            self.OPTIONAL_BILLING_FIELDS = PortalAccount.OPTIONAL_BILLING_FIELDS.copy()
+        self.OPTIONAL_BILLING_FIELDS.extend(["national_number"])
+
+    def _prepare_portal_layout_values(self):
+        values = super()._prepare_portal_layout_values()
+        national_number = request.env.user.partner_id.get_be_national_register_number()
+        values["national_number"] = national_number
+        return values
+
+    def details_form_validate(self, data):
+        error, error_message = super().details_form_validate(data)
+        national_number = data.get("national_number")
+        if national_number:
+            # The validation function expects an id_number record. We don't have
+            # that yet, so we'll mock one.
+            IdNumber = namedtuple("res.partner.id_number", ["name"])
+            id_number = IdNumber(national_number)
+            valid = not request.env[
+                "res.partner.id_category"
+            ].validate_l10n_be_national_registry_number(id_number)
```

Comment:
```suggestion
            valid = request.env[
                "subscription.request"
            ].check_be_national_register_number(national_number)
```
(this function should probably be moved from `subscription.request` to `res.partner` also. 😅 requires another pr with a major version increase for `l10n_be_cooperator_national_number`.)

Reply:
important, the proposed diff is NOT semantically equivalent. the `check` function raises an exception. the `validate` one does not.

i can do the refactoring anyway, with some try-except magic or some such.

Reply:
oh, indeed, good catch! a `try except` is indeed done in `l10n_be_cooperator_website_national_number` because of this. then maybe change the `check_be_national_register_number()` function with an argument `raise=False`? if it is `True`, it would use `validate_id_number()` and thus raise an exception if incorrect, otherwise it would call `validate_l10n_be_national_registry_number()` and return its value directly. like that, the whole namedtuple hack can stay in one place. what do you think?

Reply:
Implemented in #136.

The implementation differs a touch from your proposal: I always use `validate_id_number()`, re-raise its error if `error=True`, else catch (or don't catch) its error and return a boolean consistent with `validate_l10n_be_national_registry_number()`.

Adjusted code in this branch to assume #136 is already merged.

Reply:
good! as explained in the #136, i think the boolean value should be inverted, though.